src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1719,9 +1721,9 @@ internal ImmutableArray<DocumentId> GetRelatedDocumentIds(DocumentId documentId)
             return this.FilterDocumentIdsByLanguage(documentIds, projectState.ProjectInfo.Language);
         }
 
-        internal Solution WithNewWorkspace(Workspace workspace, int workspaceVersion)
+        internal Solution WithNewWorkspace(string? workspaceKind, int workspaceVersion, SolutionServices services)
```

Comment:
If I'm understanding this right, we're now only passing a SolutionServices object here which seems like this is friendlier than before, but we're still relying on under the covers being able to grab the workspace object? This seems a bit slimy to me: we're now passing around a supposedly safe type but still relying on the fact that we can grab the underlying workspace from it. I get other places it's perhaps an improvement but in this specific case for this specific file it feels like we're now lying to ourselves about the safety (or lack thereof).

Reply:
I can keep the `HostWorkspaceServices` parameter instead of `SolutionServices` (with some obsolete warning suppressions) if you prefer that. That said this method sets the solution services. If we remove Workspace reference from `SolutionServices` this method remains unchanged. The fact that we access workspace from the services object to implement obsolete public API is orthogonal. Maybe the name of the method should be `WithNewWorkspaceKindAndVersionAndServices` but that's a bit mouthful.

Reply:
Maybe just a comment explaining that we're also implicitly going to change Solution.Workspace with this is sufficient. I think you answered my underlying concern elsewhere -- I guess mentally I was already treating SolutionServices as "disconnected from the workspace", and even if there is that backdoor property, we just don't talk about that. But two SolutionServices objects representing different instances of the same workspace kind are not substitutable, if for no other reason that the obsolete properties still reveal the difference. But looking at the "safe" properties, they might be the same.

Reply:
> I can keep the HostWorkspaceServices parameter instead of SolutionServices (with some obsolete warning suppressions) if you prefer that.

I wouldn't mind that either, but comment seems fine.

Reply:
I'll add a comment.